dirty_cat/datasets/fetching.py

```diff
@@ -1,493 +1,396 @@
-
+        A dictionary containing:
+          - ``description``: str
+              The description of the dataset,
+              as gathered from OpenML.
+          - ``source``: str
+              The dataset's URL from OpenML.
+          - ``path``: pathlib.Path
+              The local path leading to the dataset,
+              saved as a CSV file.
+
+          The following values are added by the fetches below (fetch_*)
+          - ``read_csv_kwargs``: Dict[str, Any]
+              A list of keyword arguments that can be passed to
+              `pandas.read_csv` for reading.
+              Usually, it contains `quotechar`, `escapechar` and `na_values`.
+              See `pandas.read_csv`'s documentation for more information.
+              Use by passing `**info['read_csv_kwargs']` to `read_csv`.
+          - ``y``: str
+              The name of the target column.
 
-    References
-    ----------
-    https://data.gov.uk/dataset/road-accidents-safety-dataset
     """
 
-    return fetch_dataset(ROAD_SAFETY_CONFIG, show_progress=False)
-
-
-def fetch_medical_charge():
-    """fetches the medical charge dataset
-
-    Returns
-    -------
-    dict
-        a dictionary containing:
+    # Construct the path to the gzip file containing the details on a dataset.
+    details_gz_path = data_directory / DETAILS_DIRECTORY / f'{dataset_id}.gz'
+    features_gz_path = data_directory / FEATURES_DIRECTORY / f'{dataset_id}.gz'
+
+    if not details_gz_path.is_file() or not features_gz_path.is_file():
+        # If the details file or the features file don't exist,
+        # download the dataset.
+        warnings.warn(
+            f"Could not find the dataset {dataset_id} locally. "
+            "Downloading it from OpenML; this might take a while... "
+            "If the process is interrupted, files will be invalid/incomplete. "
+            "To fix this problem, delete the CSV file if it exists. "
+            "The system will recreate it on the next run."
+        )
+        _download_and_write_openml_dataset(dataset_id=dataset_id,
+                                           data_directory=data_directory)
+    details = _get_details(details_gz_path)
+
+    # The file ID is required because the data file is named after this ID,
+    # and not after the dataset's.
+    file_id = details.file_id
+    csv_path = data_directory / f'{details.name}.csv'
+
+    data_gz_path = data_directory / DATA_DIRECTORY / f'{file_id}.gz'
+
+    if not data_gz_path.is_file():
+        # This is a double-check.
+        # If the data file does not exist, download the dataset.
+        _download_and_write_openml_dataset(dataset_id=dataset_id,
+                                           data_directory=data_directory)
+
+    if not csv_path.is_file():
+        # If the CSV file does not exist, use the dataset
+        # downloaded by ``fetch_openml()`` to construct it.
+        features = _get_features(features_gz_path)
```

Comment:
This seems slightly suboptimal: we are calling _get_details which reads the json from the compressed file, and latter calling _get_features, which itself also reads the compressed file.

Reply:
Yes, because these are two separate files :)